src/autodiff.jl

```diff
@@ -1,16 +1,19 @@
-function autodiff(f!, initial_x::Vector)
+function autodiff(f!::Function, initial_x::AbstractArray)
```

Comment:
I think I prefer to remove the `::Function` restriction.

Sometimes I create a type to represent my objective function and then implement the `call` method on that type. 

This is really similar to a closure (_very_ similar to what Julia does under the hood for a closure -- maybe even exactly the same), but allows me to use the type elsewhere, which is convenient at times

Reply:
Sure. Would you like me to remove these restrictions also in line
https://github.com/JuliaNLSolvers/NLsolve.jl/blob/master/src/differentiable_functions.jl#L11
and at similar places?
Moreover, should it even be removed from fields in lines
https://github.com/JuliaNLSolvers/NLsolve.jl/blob/master/src/differentiable_functions.jl#L4-L6
and at similar places, and be replaced with some parameters `F1`, `F2`, `F3`? I always thought one shouldn't use fields with abstract types...

Reply:
Yes, replace them with parameters and have the parameters `<:Any` (which is implicit BTW). Fields with abstract types will not be type stable when accessing. Inplace functions are kind of an oddity here though. The access is not type stable, so the return type isn't inferred. However, since there is no return that is actually used, this type instability doesn't propagate, so the real cost ends up being minimal. On the otherhand, if it's parameterized by the function, then each function you give is a different concrete subtype and thus requires another compilation, so there's a tradeoff there. Whether specializing on the function or not is a good idea depends on the use case. I think this is a good place to use FunctionWrappers.jl sometime in the future when that all stabilizes, but for now I would just stick with their choice here.